keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,999 @@
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes decoupling the API calls for the preemption from the scheduling cycle, to enhance the scheduling throughput of the scheduling failure scenarios.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The scheduler is basically only one in a cluster,
+and hence scheduling throughput is the crutial metric for the scheduler.
+
+The scheduler schedules Pods one by one within the scheduling cycle, 
+and we basically try to reduce the API calls as much as possible to enhance the scheduling cycle throughput.
+
+The binding cycle is the example for this motivation; 
+1. The scheduling cycle decides where Pod should go to,
+2. At the end of the scheduling cycle, the scheduler reserves the Node within the scheduler's cache so that next scheduling cycle will take the current pod into consideration.
+3. The scheduling cycle ends and the binding cycle starts; the binding cycle is run asynchronously, and the scheduler starts the next scheduling cycle.
+
+This flow allows us to decouple the API call to assign Pod to the Node from the scheduling cycle so that the API call doesn't block the scheduling throughput.
+
+But, we have the similar problem with the preemption; the preemption is run at PostFilter extension point which is the part of the scheduling cycle.
+The preemption has to make some API calls to update Pods' condition and delete Pods after all, which could block the scheduling throughput.
+
+scheduler-perf [actually shows](https://github.com/kubernetes/kubernetes/blob/342da505bdefbd849b808cca3cb76c24a993025f/test/integration/scheduler_perf/config/performance-config.yaml#L641) currently the preemption scenario takes too long time, compared to others.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- The preemption plugin makes API calls for the preemption asynchronously after `PostFilter` extension point so that the scheduler can continue to other Pods' scheduling while making API calls for preemption.
+  - Until the preemption goroutine is done, the Pod won't be rescheduled.
```

Comment:
```suggestion
- Improve scheduling throughput when pods require issuing preemptions by making API calls asynchronous
```

The goal shouldn't describe the solution.

Reply:
> Until the preemption goroutine is done, the Pod won't be rescheduled.

Actually, that should not necessarily be true. If other pods finish while preemptions are being issued, maybe the pod could just go through. We could even stop the preemptions at that point, if possible.

Reply:
I think this is important. The preemptor pod doesn't know how long it will take for a preempted Pod to terminate. It can be tens of minutes (yes, I've seen customers do this).

The preemptor pod should be scheduled ASAP, as it has a high priority. It could even go to a node that it didn't nominate (it can today).

Reply:
> It can be tens of minutes

We should recommend not making such Pods lower-prioritized that could be picked up as the victims of the preemption, right? 
Regardless of whether we go with the change you're proposing here or not, selecting such pods as the preemption victims is mostly not worthwhile for scheduling; the scheduler shouldn't select those Pods in the first place.

Also, I believe no one doesn't do it currently (rather no one can do it) because if users do so with the current scheduler, the scheduler would completely stop for tens of minutes.
What do you think

Reply:
> Also, I believe no one doesn't do it currently (rather no one can do it) because if users do so with the current scheduler, the scheduler would completely stop for tens of minutes.

No, because the grace period happens after kube-scheduler has issued the DELETE API call, which is immediate.

>  Until the preemption goroutine is done, the Pod won't be rescheduled.

Although, you are only limiting this to the preemption routine, not the Pods' grace period. Re-thinking about it, that's actually similar to what we have today. Ok, if necessary, we can always adjust as we prepare for beta.